src/Components/Blazor/Validation/src/BlazorDataAnnotationsValidator.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Components.Forms
+{
+    public class BlazorDataAnnotationsValidator : ComponentBase
+    {
+        private static readonly object ValidationContextValidatorKey = new object();
+        private ValidationMessageStore _validationMessageStore;
+
+        [CascadingParameter]
+        internal EditContext EditContext { get; set; }
+
+        protected override void OnInitialized()
+        {
+            _validationMessageStore = new ValidationMessageStore(EditContext);
+
+            // Perform object-level validation (starting from the root model) on request
+            EditContext.OnValidationRequested += (sender, eventArgs) =>
+            {
+                _validationMessageStore.Clear();
+                ValidateObject(EditContext.Model);
+                EditContext.NotifyValidationStateChanged();
+            };
+
+            // Perform per-field validation on each field edit
+            EditContext.OnFieldChanged += (sender, eventArgs) =>
+                ValidateField(EditContext, _validationMessageStore, eventArgs.FieldIdentifier);
+        }
+
+        internal void ValidateObject(object value)
```

Comment:
Don't you have to keep track of the objects  you've already inspected to prevent potential infinite recursion issues?

Reply:
Yeah, you're right. Unlike MVC, the user determines what objects the validator needs to traverse, so it's much more difficult to get in to a recursion. However, you could have a trivial model like this and get in to to recursion with no other way to represent it:

```C#
public class Person
{
   [Required]
   public string Name { get; set; }

   [ValidateComplexType]
   public Person Spouse { get; set; }
}
```